File: codes/python/ch04/lebesgue_constants_zoom.py
#!/usr/bin/env python3
"""
lebesgue_constants_zoom.py

Zoomed comparison of Lebesgue constant growth for Legendre and Chebyshev nodes.

This figure omits the exponentially growing equispaced curve to allow better
visualization of the difference between:
    - Chebyshev: Λ_N = (2/π) ln(N) + O(1) ≈ O(ln N)
    - Legendre:  Λ_N ≈ O(√N)

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Études: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False  # Set to True if LaTeX is available
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
# Book color scheme
NAVY = '#142D6E'
SKY = '#7896D2'
TEAL = '#16A085'

# Output path
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch04' / 'python'
OUTPUT_FILE = OUTPUT_DIR / 'lebesgue_constants_zoom.pdf'


# -----------------------------------------------------------------------------
# Node generation
# -----------------------------------------------------------------------------
def chebyshev_nodes(N):
    """Chebyshev-Gauss-Lobatto nodes on [-1, 1]."""
    j = np.arange(N + 1)
    return np.cos(j * np.pi / N)


def legendre_nodes(N):
    """Legendre-Gauss-Lobatto nodes on [-1, 1]."""
    from numpy.polynomial import legendre as L

    if N == 0:
        return np.array([0.0])
    if N == 1:
        return np.array([-1.0, 1.0])

    # Derivative of P_N
    coeffs = np.zeros(N + 1)
    coeffs[N] = 1
    P_N = L.Legendre(coeffs)
    P_N_deriv = P_N.deriv()

    # Find roots of P'_N
    interior = P_N_deriv.roots()
    interior = np.real(interior)  # Remove tiny imaginary parts

    # Combine with endpoints
    return np.sort(np.concatenate([[-1], interior, [1]]))


# -----------------------------------------------------------------------------
# Lebesgue function computation
# -----------------------------------------------------------------------------
def lagrange_basis(x_nodes, k, x_eval):
    """Compute the k-th Lagrange basis polynomial at points x_eval."""
    n = len(x_nodes)
    x_eval = np.atleast_1d(x_eval)
    L_k = np.ones_like(x_eval, dtype=float)

    for j in range(n):
        if j != k:
            L_k *= (x_eval - x_nodes[j]) / (x_nodes[k] - x_nodes[j])

    return L_k


def lebesgue_function(x_nodes, x_eval):
    """
    Compute the Lebesgue function: Λ_N(x) = Σ_{k=0}^{N} |L_k(x)|
    """
    N = len(x_nodes) - 1
    Lambda = np.zeros_like(x_eval, dtype=float)

    for k in range(N + 1):
        Lambda += np.abs(lagrange_basis(x_nodes, k, x_eval))

    return Lambda


def lebesgue_constant(x_nodes, n_eval=2000):
    """Compute the Lebesgue constant: Λ_N = max_{x ∈ [-1,1]} Λ_N(x)"""
    x_eval = np.linspace(-1, 1, n_eval)
    Lambda = lebesgue_function(x_nodes, x_eval)
    return np.max(Lambda)


# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
def main():
    # Create figure (single panel)
    fig, ax = plt.subplots(figsize=(7, 5))

    # Compute Lebesgue constants for range of N
    N_values = np.arange(2, 51)

    Lambda_cheb_values = []
    Lambda_leg_values = []

    for N in N_values:
        Lambda_cheb_values.append(lebesgue_constant(chebyshev_nodes(N)))
        Lambda_leg_values.append(lebesgue_constant(legendre_nodes(N)))

    Lambda_cheb_values = np.array(Lambda_cheb_values)
    Lambda_leg_values = np.array(Lambda_leg_values)

    # Plot computed values
    ax.plot(N_values, Lambda_leg_values, 's-', color=TEAL,
            linewidth=1.5, markersize=5, label='Legendre')
    ax.plot(N_values, Lambda_cheb_values, '^-', color=SKY,
            linewidth=1.5, markersize=5, label='Chebyshev')

    # Plot asymptotic curves
    N_asy = np.linspace(3, 50, 100)

    # Chebyshev: (2/π) ln(N) + constant
    Lambda_cheb_asy = (2/np.pi) * np.log(N_asy) + 0.6
    ax.plot(N_asy, Lambda_cheb_asy, '--', color=SKY, alpha=0.6,
            linewidth=1.5, label=r'$(2/\pi)\ln N + 0.6$')

    # Legendre: c * sqrt(N) with fitted constant
    c_leg = Lambda_leg_values[-1] / np.sqrt(N_values[-1])
    Lambda_leg_asy = c_leg * np.sqrt(N_asy)
    ax.plot(N_asy, Lambda_leg_asy, '--', color=TEAL, alpha=0.6,
            linewidth=1.5, label=rf'${c_leg:.2f}\sqrt{{N}}$')

    ax.set_xlabel(r'$N$ (polynomial degree)')
    ax.set_ylabel(r'Lebesgue constant $\Lambda_N$')
    ax.set_title('Lebesgue Constants: Legendre vs Chebyshev', fontsize=11)
    ax.set_xlim(0, 52)
    ax.set_ylim(0, 8)
    ax.legend(loc='upper left', fontsize=10)

    # Add growth rate annotations
    ax.annotate(r'$O(\sqrt{N})$', xy=(45, Lambda_leg_values[-1]),
                xytext=(38, 6.5), fontsize=11, color=TEAL,
                arrowprops=dict(arrowstyle='->', color=TEAL, lw=0.8))
    ax.annotate(r'$O(\ln N)$', xy=(45, Lambda_cheb_values[-1]),
                xytext=(38, 2.0), fontsize=11, color=SKY,
                arrowprops=dict(arrowstyle='->', color=SKY, lw=0.8))

    # Clean styling
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.spines['left'].set_color('#444444')
    ax.spines['bottom'].set_color('#444444')
    ax.tick_params(colors='#444444')
    ax.yaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
    ax.xaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
    ax.set_axisbelow(True)

    plt.tight_layout()

    # Save figure
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    fig.savefig(OUTPUT_FILE, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(OUTPUT_FILE.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {OUTPUT_FILE.resolve()}')

    # Print comparison table
    print("\nLebesgue Constants Comparison (Legendre vs Chebyshev):")
    print("-" * 50)
    print(f"{'N':>4} {'Legendre':>15} {'Chebyshev':>15} {'Ratio':>10}")
    print("-" * 50)
    for i, N in enumerate(N_values[::10]):  # Print every 10th value
        idx = i * 10
        ratio = Lambda_leg_values[idx] / Lambda_cheb_values[idx]
        print(f"{N:>4d} {Lambda_leg_values[idx]:>15.4f} "
              f"{Lambda_cheb_values[idx]:>15.4f} {ratio:>10.2f}")
    print("-" * 50)

    plt.close(fig)


if __name__ == '__main__':
    main()

File: codes/julia/ch04/lebesgue_constants_zoom.jl
# lebesgue_constants_zoom.jl
#
# Zoomed comparison of Lebesgue constant growth for Legendre and Chebyshev nodes.
#
# This figure omits the exponentially growing equispaced curve to allow better
# visualization of the difference between:
#     - Chebyshev: Lambda_N = (2/pi) ln(N) + O(1)
#     - Legendre:  Lambda_N ~ O(sqrt(N))
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LaTeXStrings
using Polynomials

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
# Book color scheme
const NAVY = colorant"#142D6E"
const SKY  = colorant"#7896D2"
const TEAL = colorant"#16A085"

# Output path
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch04", "julia")

# -----------------------------------------------------------------------------
# Node generation
# -----------------------------------------------------------------------------
chebyshev_nodes(N) = [cos(j * π / N) for j in 0:N]

function legendre_nodes(N)
    N == 0 && return [0.0]
    N == 1 && return [-1.0, 1.0]

    # Build P_N via recurrence
    P_prev = [1.0]
    P_curr = [0.0, 1.0]

    for n in 1:N-1
        P_next = ((2n + 1) .* [0; P_curr] .- n .* [P_prev; zeros(2)]) ./ (n + 1)
        P_prev = P_curr
        P_curr = P_next
    end

    P_N_poly = Polynomial(P_curr)
    P_N_deriv = derivative(P_N_poly)

    interior = roots(P_N_deriv)
    interior = real.(filter(r -> abs(imag(r)) < 1e-10, interior))

    return sort(vcat([-1.0], interior, [1.0]))
end

# -----------------------------------------------------------------------------
# Lebesgue function computation
# -----------------------------------------------------------------------------
function lagrange_basis(x_nodes, k, x_eval)
    n = length(x_nodes)
    L_k = ones(length(x_eval))
    for j in 1:n
        if j != k
            L_k .*= (x_eval .- x_nodes[j]) ./ (x_nodes[k] - x_nodes[j])
        end
    end
    return L_k
end

function lebesgue_function(x_nodes, x_eval)
    N = length(x_nodes) - 1
    Lambda = zeros(length(x_eval))
    for k in 1:N+1
        Lambda .+= abs.(lagrange_basis(x_nodes, k, x_eval))
    end
    return Lambda
end

function lebesgue_constant(x_nodes; n_eval=2000)
    x_eval = collect(range(-1, 1, length=n_eval))
    Lambda = lebesgue_function(x_nodes, x_eval)
    return maximum(Lambda)
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Create figure (single panel)
    fig = Figure(size = (700, 500))
    ax = Axis(fig[1, 1],
              xlabel = L"N" * " (polynomial degree)",
              ylabel = L"\Lambda_N" * " (Lebesgue constant)",
              title = "Lebesgue Constants: Legendre vs Chebyshev",
              limits = ((0, 52), (0, 8)))

    # Compute Lebesgue constants for range of N
    N_values = 2:50

    Lambda_cheb_values = [lebesgue_constant(chebyshev_nodes(n)) for n in N_values]
    Lambda_leg_values = [lebesgue_constant(legendre_nodes(n)) for n in N_values]

    # Plot computed values
    scatterlines!(ax, collect(N_values), Lambda_leg_values,
                  color = TEAL, linewidth = 1.5, markersize = 8, marker = :rect,
                  label = "Legendre")
    scatterlines!(ax, collect(N_values), Lambda_cheb_values,
                  color = SKY, linewidth = 1.5, markersize = 8, marker = :utriangle,
                  label = "Chebyshev")

    # Plot asymptotic curves
    N_asy = collect(range(3, 50, length=100))

    # Chebyshev: (2/pi) ln(N) + constant
    Lambda_cheb_asy = (2/π) .* log.(N_asy) .+ 0.6
    lines!(ax, N_asy, Lambda_cheb_asy, linestyle = :dash, color = SKY,
           linewidth = 1.5, alpha = 0.6, label = L"(2/\pi)\ln N + 0.6")

    # Legendre: c * sqrt(N) with fitted constant
    c_leg = Lambda_leg_values[end] / sqrt(N_values[end])
    Lambda_leg_asy = c_leg .* sqrt.(N_asy)
    lines!(ax, N_asy, Lambda_leg_asy, linestyle = :dash, color = TEAL,
           linewidth = 1.5, alpha = 0.6,
           label = latexstring("$(round(c_leg, digits=2))\\sqrt{N}"))

    # Growth rate annotations with arrows
    text!(ax, 38.0, 6.5, text = L"O(\sqrt{N})", fontsize = 11, color = TEAL)
    text!(ax, 38.0, 2.0, text = L"O(\ln N)", fontsize = 11, color = SKY)

    axislegend(ax, position = :lt, framevisible = false, labelsize = 10)

    # Save figure
    mkpath(OUTPUT_DIR)
    save(joinpath(OUTPUT_DIR, "lebesgue_constants_zoom.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "lebesgue_constants_zoom.png"), fig, px_per_unit = 4)

    println("Figure saved to: ", joinpath(OUTPUT_DIR, "lebesgue_constants_zoom.pdf"))

    # Print comparison table
    println("\nLebesgue Constants Comparison (Legendre vs Chebyshev):")
    println("-" ^ 50)
    println("$(rpad("N", 4)) $(lpad("Legendre", 15)) $(lpad("Chebyshev", 15)) $(lpad("Ratio", 10))")
    println("-" ^ 50)
    for i in 1:10:length(N_values)
        n = N_values[i]
        ratio = Lambda_leg_values[i] / Lambda_cheb_values[i]
        println("$(lpad(n, 4)) $(lpad(round(Lambda_leg_values[i], digits=4), 15)) $(lpad(round(Lambda_cheb_values[i], digits=4), 15)) $(lpad(round(ratio, digits=2), 10))")
    end
    println("-" ^ 50)

    return fig
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
